examples/Conversational_agents_using_tools.ipynb

```diff
@@ -0,0 +1,1095 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "30995a82",
+   "metadata": {},
+   "source": [
+    "# Conversational Agents using Tools\n",
+    "\n",
+    "This notebook takes you through how to use LangChain to augment an OpenAI model with access to external tools. In particular, you'll be able to create LLM agents that use custom tools to answer user queries.\n",
+    "\n",
+    "\n",
+    "## What is Langchain\n",
+    "\n",
+    "LangChain is a [framework](https://python.langchain.com/en/latest/index.html) for developing applications powered by language models. Its main features include prompt chaining (running a sequence of language model queries in which outputs of earlier queries are made part of the input of later queries) and utilities for converting language models into tool-using agents.\n",
+    "\n",
+    "\n",
+    "## Why do LLMs need to use Tools\n",
+    "\n",
+    "Providing LLMs access to tools can enable them to answer questions with context directly from search engines, APIs or your own databases. Instead of answering directly, an LLM with access to tools can perform intermediate steps to gather relevant information. Tools can also be used in combination. [For example](https://python.langchain.com/en/latest/modules/agents/agents/examples/mrkl_chat.html), a language model can be made to use a search tool to lookup quantitative information and a calculator to execute calculations.\n",
+    "\n",
+    "## Notebook Sections\n",
+    "\n",
+    "- **Setup:** Import packages and connect to a Pinecone vector database.\n",
+    "- **LLM Agent:** Build an agent that leverages a modified version of the [ReAct](https://react-lm.github.io/) framework to do chain-of-thought reasoning.\n",
+    "- **LLM Agent with History:** Provide the LLM with access to previous steps in the conversation.\n",
+    "- **Knowledge Base:** Create a knowledge base of \"Stuff You Should Know\" podcast episodes, to be accessed through a tool.\n",
+    "- **LLM Agent with Tools:** Extend the agent with access to multiple tools and test that it uses both to answer questions."
```

Comment:
stumbled a bit when I ran across the word "both". Are there going to be two tools or more tools? Either way, recommended replacing "both" with "them" just to cover cases where there are more than two tools, or cases where they aren't being used at the same time.

Reply:
Done